modules/ui/fields/date.js

```diff
@@ -0,0 +1,268 @@
+    });
+
+    let monthCombo = uiCombobox(context, 'date-month')
+        .data(monthNames.map((monthName, i) => {
+            return {
+                id: i + 1,
+                value: monthName,
+                title: monthName,
+                display: selection => selection.append('span')
+                    .attr('class', 'localized-text')
+                    .text(monthName),
+                terms: alternativeMonthNames[i],
+            };
+        }));
+
+    function date(selection) {
+
+        var wrap = selection.selectAll('.form-field-input-wrap')
+            .data([0]);
+
+        wrap = wrap.enter()
+            .append('div')
+            .attr('class', 'form-field-input-wrap form-field-input-' + field.type)
+            .merge(wrap);
+
+        yearInput = wrap.selectAll('input.date-year')
+            .data([0]);
+        eraInput = wrap.selectAll('input.date-era')
+            .data([0]);
+        monthInput = wrap.selectAll('input.date-month')
+            .data([0]);
+        dayInput = wrap.selectAll('input.date-day')
+            .data([0]);
+
+        formatParts.forEach(part => {
+            switch (part) {
+                case 'year':
+                    yearInput = yearInput.enter()
+                        .append('input')
+                        .attr('type', 'number')
+                        .attr('class', 'date-year')
+                        .attr('id', field.domId)
+                        .call(utilNoAuto)
+                        .merge(yearInput);
+                    break;
+                case 'era':
+                    eraInput = eraInput.enter()
+                        .append('input')
+                        .attr('type', 'text')
+                        .attr('class', 'date-era')
+                        .call(eraCombo)
+                        .merge(eraInput);
+                    break;
+                case 'month':
+                    monthInput = monthInput.enter()
+                        .append('input')
+                        .attr('type', 'text')
+                        .attr('class', 'date-month')
+                        .call(monthCombo)
+                        .merge(monthInput);
+                    break;
+                case 'day':
+                    dayInput = dayInput.enter()
+                        .append('input')
+                        .attr('type', 'number')
+                        .attr('class', 'date-day')
+                        .call(utilNoAuto)
+                        .merge(dayInput);
+                    break;
+            }
+        });
+
+        yearInput
+            .on('change', change)
+            .on('blur', change);
+        eraInput
+            .on('blur', change)
+            .on('change', change);
+        monthInput
+            .on('blur', change)
+            .on('change', change);
+        dayInput
+            .on('change', change)
+            .on('blur', change);
+    }
+
+
+    function change() {
```

Comment:
The fields don’t seem to be updating when undoing and redoing.

Reply:
I noticed this as well when testing locally. Maybe some change is needed in https://github.com/OpenHistoricalMap/iD/blob/518546cc5d4331e148a1718e66d44adf6bb78079/modules/core/history.js? But no idea what change! Do you think the improved date entry experience is worth merging and deploying this without undo working (and then fix undo separately)? I could see arguments either way.

Reply:
Fixed in ab49980fdc93c61264f0b04d62e0321b6f729ff0.

Reply:
In my testing, now the numeric element (day) does clear on undo, and in cases where there's a value for month and I change it, the month reverts (as expected). But in cases where month is blank and I add a value, the value stays after I hit undo.

Here's a screen cap of that https://recordit.co/OT5NNxLWoA

Reply:
Fixed in 0671357a4c4cff014eb877c0e1c4f22ec297587d.